src/server/builder.ts

```diff
@@ -319,18 +324,19 @@ namespace ts.server {
         getFilesAffectedBy(scriptInfo: ScriptInfo): string[] {
             this.ensureProjectDependencyGraphUpToDate();
 
+            const singleFileResult = scriptInfo.hasMixedContent ? [] : [scriptInfo.fileName];
             const fileInfo = this.getFileInfo(scriptInfo.path);
             if (!fileInfo || !fileInfo.updateShapeSignature()) {
-                return [scriptInfo.fileName];
+                return singleFileResult;
             }
 
             if (!fileInfo.isExternalModuleOrHasOnlyAmbientExternalModules()) {
-                return this.project.getFileNamesWithoutDefaultLib();
+                return map(filter(this.project.getScriptInfosWithoutDefaultLib(), shouldEmitFile), info => info.fileName);

```

Comment:
this chunk of code look awfully similar to the one above. Can you pull this into method in base class? Also can you replace pair `map` + `filter` with loop to avoid creating intermediate array?